great_expectations/checkpoint/checkpoint.py

```diff
@@ -256,23 +263,32 @@ def _sort_actions(self) -> List[CheckpointAction]:
     def save(self) -> None:
         from great_expectations.data_context.data_context.context_factory import project_manager
 
-        store = project_manager.get_checkpoints_store()
-        key = store.get_key(name=self.name, id=self.id)
-
-        store.update(key=key, value=self)
-
-    def _add_to_store(self) -> None:
-        """This is used to persist a checkpoint before we run it.
-
-        We need to persist a checkpoint before it can be run. If user calls runs but hasn't
-        persisted it we add it for them.
-        """
-        from great_expectations.data_context.data_context.context_factory import project_manager
+        for validation in self.validation_definitions:
+            try:
+                validation.save()
+            except (
+                ValidationDefinitionSaveError,
+                ValidationDefinitionRelatedResourceSaveError,
+            ) as e:
+                raise CheckpointRelatedResourceSaveError(
+                    name=self.name,
+                    related_resource_type=ValidationDefinition.__class__.__name__,
+                    related_resource_name=validation.name,
+                ) from e
 
         store = project_manager.get_checkpoints_store()
         key = store.get_key(name=self.name, id=self.id)
 
-        store.add(key=key, value=self)
+        try:
+            if store.has_key(key):
+                store.update(key=key, value=self)
+            else:
+                store.add(key=key, value=self)
+        except (
+            StoreBackendError,  # Generic error raised by store backends (namely GXCloudStoreBackend) # noqa: E501
+            ValueError,  # Generic error by certain stores/store backends
+        ) as e:
```

Comment:
I hate having to do this but different stores and store backends raise different things and we don't have uniformity around error handling or types